### ISSUE TITLE
AltTab's window sometimes "hangs" for 0.5 to 1.5 sec after choosing a window or cancelling

### ISSUE THREAD

**Issue by mfn** [lwouis/alt-tab-macos#563]
Version: 6.0.0
**Describe the bug**
I did not notice this with the version before 6 I was using and I just upgraded.

I would say 1 out of 5 times, when I choose (keyboard or mouse) or not choose to select a window (ESC), the over of the thumbs hangs around a bit longer before it disappears (i.e. it's not instant).

Hard to properly _reproduce_, it just happens sometimes.

- _Maybe_ it's after some time _not_ bringing up AltTab and then quickly choosing a window. That is what just happened while I wrote the issue and quickly wanted to switch somewhere else.
- Yep, same happened again. I finished above sentence and _then_ I could immediately reproduce it again

I use "RecordIT" and tried to capture it, but it doesn't properly reflect the behaviour due to low FPS :/

----
![jbETlzsvkC](https://user-images.githubusercontent.com/87493/91526906-fc48ec80-e904-11ea-97d0-beb8631c5c5e.gif)

**Comment by mfn** [comment]
I see everyone using loom or so, but it needs registering AFAICS which I didn't want to use. I use the free version of RecordIT (due to low FPS) but that's _usually_ enough for me.

If anyone can recommend another screen recording app which doesn't need registering and produces better results, I'm open to suggestions!

**Comment by lwouis** [comment]
Hey again @mfn! Thanks for sharing this issue!


I noticed this during development for 2 days. I thought I was going crazy because it's so inconsistent like why is it happening? I tried to capture it in Instruments to profile it, but it only happened once there, and there was basically no activity in AltTab, so I thought "ok it's the WindowServer being slow to respond for some reason, nothing I can do about that".

It's either some multi-threading issue, or OS interaction like stressing out the WindowServer in burst or something. I'm sad to hear it's happening on other people's computer too. I thought it was only on my system. I'll try to work at it, somehow.

For the video software. I used OBS in the past. It records in top quality, but then you have the issue of where to host the content to share it. Loom is incredibly plug-and-play. Yes you need an account, but then it's a one-click to record and share with others. They pay the video hosting, so I feel it's fair to sign up. They are trying to run a business, and their product is good. Any free software is going to put the burden of hosting the video on you.

**Comment by lwouis** [comment]
Questions to help me investigate:

* Do you remember these freezes happening on previous releases, or is it a new thing?
* Do you have BetterTouchTool running?

**Comment by lwouis** [comment]
I got a ~1s freeze during profiling. It's always during shortcut release, when the AltTab main window should hide, that it freezes before hiding.

![image](https://user-images.githubusercontent.com/106195/91536844-558a3d80-e950-11ea-979c-253664087d09.png)
![image](https://user-images.githubusercontent.com/106195/91536850-57540100-e950-11ea-840e-7aa10b40e5da.png)

When I look at these, especially the first one, I see the main thread is mostly idling. Basically it's running its code very efficiently, then doing nothing. If the freeze is not due to some code in the main thread not finishing fast, then the only explanation I have is that it's the window server being slow giving the thumbnails or the AX info (e.g. titles, roles, etc).

What puzzles me, is why this is happening on the latest version. If anything, the code is more efficient than ever. Anybody has an idea?

**Comment by mfn** [comment]
> Do you remember these freezes happening on previous releases, or is it a new thing?

No, it started today when I installed 6.x, as I wrote:
>  I did not notice this with the version before 6 I was using and I just upgraded.

> Do you have BetterTouchTool running?

No, never heard of. 

I might even downgrade, it really happens very often; now I would say the number is 3 out of 5 for me.

**Comment by lwouis** [comment]
I've been trying to reproduce for 2 days now, but I can't get it to happen anymore. I was running some local changes that are now release as part of v6.1.0.

@mfn could you please tell me if the issue still happens to you with v6.1.0? I'm really trying hard like plugin my 4K TV and putting 8 4k youtube video to stress the WindowServer, and other things, but I can't reproduce the issue for the life of me. Maybe of the changes in 6.1.0 got rid of the issue. I hope so at least.

**Comment by mfn** [comment]
Roger, I jut updated to v6.1.0 and will continue observing the situation!

**Comment by frypf** [comment]
I updated earlier today and have unfortunately noticed this a few times since. I hadn't actually noticed it in previous versions, although I've been having another issue and it's only been while fiddling with the AltTab settings that it's even come up.

I've also found it near impossible to reproduce. For me it seems to be related to changing the fade out animation and apparition delay, and I've only had it happen while the prefs window is open. Even then it's not consistent, happening maybe three times out of probably hundreds of app switches over the last few hours.

**Comment by lwouis** [comment]
> I've only had it happen while the prefs window is open

I'm not sure 100% but i think it is also the case for me that whenever i witnessed this issue, the prefs window was open. Same for you @mfn?

**Comment by mfn** [comment]
> For me it seems to be related to changing the fade out animation and apparition delay, and I've only had it happen while the prefs window is open

I can't confirm this, I don't remember the last time I was in the settings [1] and they're never open.

But besides that, since 6.1.0 I have not experiences the delay _yet_.

[1] Actually I do, when I learned that we can switch in-app, I enabled it a few weeks ago ;)

**Comment by mfn** [comment]
The delay is back AKA today, I just started working ~10mins ago, I already noticed in twice. can't say much about yesterday, was too late & tired.

**Comment by lwouis** [comment]
Anybody good with Instruments? I'm kind of stuck here on how to investigate this issue further.

**Comment by mfn** [comment]
It happened a few seconds ago and what I was doing _right before_: I switched a branch in PhpStorm and re-indexed was triggered (lots of files changed), and during that time I switched.

CPU spike/lag/throttle/… ? However I could not reproduce it that way. Maybe because the second time the disk cache softens this… 🤷‍♀️

**Comment by mfn** [comment]
No, that's not it. At least: not alone.

I had it a few times happen in the last 2 hours without being aware of triggering an CPU intensive operation before that. E.g. switching already before loaded browser windows.

**Comment by lwouis** [comment]
My guess is that it's either:

* A concurrency issue within AltTab. Some wrong code undeterministically behaves badly, depending on execution speed of each thread, which is random.
* A bottleneck on WindowServer. Perhaps AltTab is too efficient and asks for too much data, too fast, and saturates WindowServer, which then struggles to clean AltTab own window from the screen. WindowServer communication patterns are undeterministic
* A resources bottleneck. Maybe the screenshots require too much RAM access, and depending on access patterns which are undeterministic, we get a freeze.

**Comment by mfn** [comment]
I mentioned it elsewhere but not here yet, so to have it one place my specs:
- 10.15.6
- MacBook Pro (16-inch, 2019)
- 2,4 GHz 8-Core Intel Core i9
- 64 GB 2667 MHz DDR4
- APPLE SSD AP0512N
- AMD Radeon Pro 5300M 4 GB
  Intel UHD Graphics 630 1536 MB

Additionally I use two external FullHD monitors.

From what I remember, the only other software constantly "active" regarding the display is f.lux (disabled Apples own version of that) but I don't remember it ever interfering with anything.

**Comment by frypf** [comment]
Not sure I can help much further with this one unfortunately. My machine is considerably less powerful (mac mini / i3 processor / 16gb ram), so in theory the proposed resource bottleneck might be a more frequent problem for me, but then I'm only running a single screen at a much lower res and not generally doing anything intensive. Other times I've experienced general lag on my system it's usually been down to WindowServer.

I've only noticed the delay once more since my previous post. The prefs window was definitely open again at the time, but I'd been changing a different setting so I don't think it's related to changing the value of any specific setting (for me at least). I will of course keep monitoring and report back if I notice any other discernible pattern as to it's cause.

**Comment by lwouis** [comment]
@frypf I also suspected it was an interaction with the Preferences window, but @mfn says it's never open on his machine yet he experiences the issue.

From my perspective:

* Issue happens randomly if I try for 30min
* I captured the issue during Instruments profiling, yet I don't see any usable information on the UI (see screenshot above)
* No clue about root cause
* No clue how to reproduce deterministically
* Even the symptoms are inconsistent

**Comment by frypf** [comment]
@lwouis I was just adding to my previous comment when your reply came through. I've managed to narrow it down slightly more, but no idea how helpful this may be:

I've been able to reproduce it a few more times in the last hour. While the following steps won't reproduce the problem every time (more like 1 in 20), I've not noticed it at all under any other circumstances:

- Open AltTab prefs window
- Interact with a slider based preference
- Switch to another app
- Switch back to prefs window
- Delay happens. Occasionally. But only maybe ~0.5 - 1s, not the longer delay that @mfn mentions.

I also noticed that a few times the prefs window takes a similar amount of time to appear when I called it via the menu bar, rather than appearing near-instantly as it usually does - although this seems to have no bearing on whether I then experience the same delay when switching back to AltTab from another app.

At this point I'm increasingly unsure that my own issue isn't just down to resources and WindowServer, although it's weird that it only seems to happen after I open the prefs anew and adjust a slider, and even then it's only occasional. Subsequent interactions with the sliders don't seem to reproduce it, only when I change one after reopening the prefs. Being so specific and intermittent, it's certainly not affecting my workflow to any extent as it seems to for @mfn. Perplexing, and quite possibly unrelated / unhelpful with the original problem, so apologies if my input has been a red herring!

**Comment by lwouis** [comment]
@frypf doing these exact steps seems to make it more likely to happen indeed. Thank you a lot for your support in attempting to find reproduction steps. I really appreciate it!

**Comment by frypf** [comment]
No worries, glad it's not just an idiosyncrasy of my own mac!

**Comment by mfn** [comment]
I'm not sure if I did not pay proper attention or not so far, but what I noticed definitely since yesterday day is that it takes longer than expected to "bring up" the list of windows.

I would say 1 out of 10 times I press cmd-tab and it takes up to 0.25-0.5s, hard to pinpoint) to _bring up_ the window.

I wonder if I should go back to 5.x and positively confirm it does not happen there? 🤔 As I initially wrote, I didn't remember having this before

**Comment by lwouis** [comment]
@mfn yes of course, any data point is helpful in trying to find the root cause so we can fix it. Comparing with v5, sharing screen capture of the symptoms, etc, it's all very useful, as these problems seem to vary across machines

**Comment by mfn** [comment]
Roger, [using 5.3.0 now](https://github.com/lwouis/alt-tab-macos/releases/tag/v5.3.0)

**Comment by lwouis** [comment]
Could you please [try out this build I made for another ticket](https://github.com/lwouis/alt-tab-macos/issues/566#issuecomment-686715851)? It's performing great on my system.

**Comment by mfn** [comment]
> ![image](https://user-images.githubusercontent.com/87493/92204999-0464d700-ee85-11ea-879c-d453d1a9604d.png)

I don't recall having this hangs since I'm on v5. I'll stay on that version and continue throughout today.

**Comment by lwouis** [comment]
[New build](https://github.com/lwouis/alt-tab-macos/files/5173939/AltTab.app.zip). I changed some displays related things, while working on #566. Maybe it helps? I haven't seen the lag yet

**Comment by lwouis** [comment]
I caught the issue on video, in a profiling session. I go through my analysis, and why i'm stuck.

If you can, please [check out the video and let me know if you have an idea](https://www.loom.com/share/b278daf6b2664b9cafb049ab6d6c1f8e) what could be going on.

[Here is the `.trace` document](https://github.com/lwouis/alt-tab-macos/files/5174471/Instruments.trace.zip) if you want to study it in Instruments for yourself

**Comment by lwouis** [comment]
Oh wow, I reviewed the footage, and if actually looks like the whole system UI is freezing up for a second. I first noticed that the Instruments moving graph was freezing, but then I thought that it may be freezing because AltTab itself is freezing or something. They are interacting somehow so it's not enough to conclude.

But then I checked the iStats graphs on the top right corner of the screen. Pay attention to the SSD and NET graphs on the right. Notice how one refreshes, then the other, then the first, etc. Count the tic-toc-tic-toc-... in your head. When AltTab freezes, this part of the UI also freezes.

I really looks like macOS is the culprit here! What do you think?

**Comment by lwouis** [comment]
I'm not able to reproduce the lag (for now at least) using this build. Could you please [try it out and let me know](https://github.com/lwouis/alt-tab-macos/files/5174594/AltTab.app.zip)?

Btw I'm sorry to ask for help on so many builds, but it's kind of the only way to do QA in the real world scenarios we have here. It's beta-testing basically. Your support is really appreciated 👍

**Comment by frypf** [comment]
> It's beta-testing basically. Your support is really appreciated 👍

Absolutely no problem, you're putting in the hard work pushing out a bunch of simultaneous solutions to issues and feature enhancements, so it's kind of the least we can do. At least you're up front about it, which is more than I can say for several bits of paid software I've had in the past.
I'm starting to confuse myself a bit referring to all the various builds between threads though so we need a naming convention for these tests. Maybe just use the github file number for now as a quick/ dirty solution, or reference the latest commit you've incorporated when you write the link?

I've noticed the lag a couple of times with the version above, but still only with the prefs window open, so not sure if it's a slightly different issue to @mfn. I've written my other findings in our discussion #566, so I'll avoid polluting this thread any further unless I start noticing this without the prefs being open.

**Comment by lwouis** [comment]
> Maybe just use the github file number for now as a quick/ dirty solution, or reference the latest commit you've incorporated when you write the link?

You can also simply link to my link. For example you would say "I tested and had results X with [this build](https://github.com/lwouis/alt-tab-macos/issues/563#issuecomment-687093108)". Notice how the link point to my comment above, itself containing the build. That way it's easy to know because you click the link, and see the build, in its historical context.

**Comment by SoPat712** [comment]
Shouldn't this be open if the bug still exists?

**Comment by lwouis** [comment]
Oh, it was closed automatically because of a reference in a commit, which I forgot to remove. My bad, reopening.

If you want to help fix the issue, you can [review the latest investigation from this comment](https://github.com/lwouis/alt-tab-macos/issues/563#issuecomment-687085023), or you can try to help us understand the root cause by finding scenarios that deterministically reproduce the problem, or guess at what could be at play here.

**Comment by SoPat712** [comment]
Are we sure it's not display stuttering or something else unfixable? Or heavy processor load as a result of alt-tab causing a drop in frames?

**Comment by lwouis** [comment]
These questions have been answered in the discussion above. Please refer to the link I just shared. It contains a video in which I showcase the issue, and go over my best guess as to what is happening. Yes it could be a system limitation. It could also be an issue with AltTab. Hard to say. See the video. You can also use the `.trace` and import it in Instruments, to try to understand the issue.

I think this was introduce with v6 (previous version was 5.3.0). The main change in v6 is that thumbnails are retrieved asynchronously compared to synchronously previously. [This is the main commit](da16a0b8f9ef02855e38eef5538eaaf49c8c5861). This change may have introduce a bug, or may have unveiled limitations of the OS.

**Comment by lwouis** [comment]
> like thread.sleep for a couple of milliseconds to let the OS catch up

What would be the point of that? How long would we wait? How would we know that waiting is need? I don't see the point of this. We could go back to doing it synchronously if we want to wait. Also we are not sure at all the OS needs to catch up. It needs to be proved.

> is there a way we can cache unchanging thumbnails in the background so that alt-tab doesn't have to work so hard

This is exactly what v6 introduced: the thumbnails are shown immediately within a less than 50ms from the shortcut trigger. Then AltTab asynchronously update them 1 by 1. In the past, if you had 100 open windows, AltTab would synchronously get 100 screenshots from the OS (in parallel tasks, with a fork/join approach, so it was not too slow). With v6, AltTab will get the first 3 (most looked at) window screenshots synchrously, then show the UI with the other 97 thumbnails in their previous/cached version. Then start updating them 1 by 1. If the user closes the UI during this fetching, it stops, so I don't know if the fetching is even what's causing the freeze. I don't know what's causing the freeze at all.

Have you watched the video I recommended above?

**Comment by SoPat712** [comment]
I've just watched the video above, and I think I understand the issue a lot better now. Regarding sleep, thread().sleep in java works to stutter certain processes that take place on processor cores, setting them in an artificial suspended state,. It gives time for alt-tab or other applications to get its own processing time. This is from Oracle documentation "This is an efficient means of making processor time available to the other threads of an application or other applications that might be running on a computer system". I'm not sure how this works in Swift, and don't know how it would really work, and once again, I'm not a MacOS app developer, so make sure to take my advice with a gallon of salt haha.

**Comment by mfn** [comment]
> Maybe just use the github file number for now as a quick/ dirty solution, or reference the latest commit you've incorporated when you write the link?

I also have the problem that, after a few days, I literally don't remember which version… so far it's manageable but maybe error prone 😅

From https://github.com/lwouis/alt-tab-macos/issues/563#issuecomment-686930623 
- I have been on v5 since that time and worked a bit over the last days, not once do I recall a situation where this lag happened.

I'll now use the build from https://github.com/lwouis/alt-tab-macos/issues/563#issuecomment-687093108

**Comment by lwouis** [comment]
@mfn please use the public release instead. All the builds in this thread were made while I was trying to close the latest releases. The latest and greatest code is now in 6.2.0, and it would be best to use it as the baseline. If we come up with hypothesis I'll make new test builds, but for now let's only discuss the latest public release for the issue, and 5.3.0, 6.0.0 to confirm if indeed the issue started with 6.0.0

**Comment by mfn** [comment]
Ok, done!

**Comment by mfn** [comment]
Oh my, I just re-read the title I wrote:
> AltTab list of thumbs sometimes hangs for 0.5 to 1.5 sec after choosing a window or cancelling

When in fact in all recent tests (i.e at _minimum_ since last week) what I experienced was rather a delay in **showing** AltTab windows at all

Aka, this delay I was all the time recently saying I still have it, in fact it was about bringing up AltTab. And not so much after choosing a window and then switch.

----
I am so so sorry and I start to confuse myself here. AKA the worst possible combination of all when reporting and then not even the same thing.

I will swear I will pay more attention to this.

----
I was just about to say "I still can reproduce it" when in fact I was really talk about the "startup" delay I experience sometimes when I press CMD-tab to bring up the windows.

I definitely did do experience the original issue but at some point this all got fuzzy and I started to report I experience the original issue when in fact I experiences the one I'm describing now.

----
I need to go back and collect myself, I really really feel back for this confusion.

Again, apologies and please let me know if you have an idea how to best proceed with _this_ issue.

☹️

**Comment by frypf** [comment]
> I'm not sure if I did not pay proper attention or not so far, but what I noticed definitely since yesterday day is that it takes longer than expected to "bring up" the list of windows.

> When in fact in all recent tests (i.e at minimum since last week) what I experienced was rather a delay in showing AltTab windows at all

I was reading through the thread again and was actually going to ask for clarification on this. Both these quotes describe the issue I'm now noticing more myself, which seems to have moved away from the original delay upon releasing the shortcut (as mentioned in both the title and shown in @lwouis's [video](https://github.com/lwouis/alt-tab-macos/issues/563#issuecomment-687085023)).

I'm experiencing an infrequent delay in showing either the AltTab switcher _or_ prefs, but have not noticed any more delays when actually switching to another app (as in the titled issue). I had just been attributing this to resources and WindowServer, or possibly my own key-release timing during space transitions after the work @lwouis put in on #566. However I've since noticed it occasionally without any space transition being involved, for example this is how it looks when showing the prefs:

![prefsDelayGlitch](https://user-images.githubusercontent.com/65832791/92392745-a818e600-f116-11ea-80c3-b2b607ef8967.gif)

When switching apps without a space transition, it doesn't typically manifest with the visual glitch and so is much harder to capture. If a space transition is involved it looks like this:

![switcherDelayGlitch](https://user-images.githubusercontent.com/65832791/92396351-cf72b180-f11c-11ea-8a53-8ae745689bbd.gif)

Just want to check with both @lwouis and @mfn that I'm not further confusing the issue or talking at cross purposes. I'll keep searching for reproduction steps, but it's proving really hard to put a finger on.

**Comment by lwouis** [comment]
Anyone knows a good macOS developer we could ask for advice? These graphical glitches are beyond me 🙉 

I don't know what to do. Everyone seems to be experiencing slightly different, indeterministic symptoms. And these symptoms have no trace in profiling tools.

**Comment by andylima** [comment]
> Anyone knows a good macOS developer we could ask for advice?

<http://twitter.com/MacSparky> knows a lot of Mac devs — I think he could be of help. 🙂

(I'm also having this issue since using that "disable thumbnails" preview build…)

**Comment by frypf** [comment]
I think in one of your videos you mentioned that counterintuitively the rendering glitches may be caused by AltTab's windows attempting to show too early, before they're actually ready. This tallies with the fact that if I turn the "Apparition delay" up (above ~500ms) I don't remember seeing the graphical glitches happen at all for same-space transitions. Interestingly though, even with this set high I still experience occasional similar glitches when opening the prefs (which obviously isn't subject to the Apparition delay setting). When switching apps across spaces the issue also still happens, albeit less frequently (the switcher window sometimes ignores that setting entirely and attempts to show itself, sometimes successfully, sometimes with the glitch).

Your knowledge is levels above mine so I feel like I only ever end up highlighting issues without being able to contribute anything concrete to resolving them - hence my possibly overzealous hunt for reproduction steps.

I've only been a lurker on the Yabai forum, but @koekeishiya always seems a fountain of knowledge, although having read an increasing number of threads both here and there I know you guys already have a connection, and he's never not busy..

**Comment by lwouis** [comment]
Oh! I was working on #289 and I caught this in the profiler:

![image](https://user-images.githubusercontent.com/106195/92408895-cb28b180-f179-11ea-989b-96402ac9160d.png)

This is a big catch! In terms of timing, I have been personally suspicious that some of the freezes started happening around the time I introduced the app badges. It's a synchronous request I added kind of quickly. It's *very* possible, as the picture shows, that this call sometimes hangs a lot, when the Dock or WindowServer is busy.

I will try to make the call async like the thumbnails

**Comment by lwouis** [comment]
Here is a [build with async app badges](https://github.com/lwouis/alt-tab-macos/files/5184551/AltTab.app.zip). Could you please tell me if it improved things?

**Comment by mfn** [comment]
I just installed it.

After a few minutes, when I switched windows _within_ PhpStorm, upon <kbd>CMD-~</kbd> it took a few more ms than expected until the list of windows did show up (5 windows in total).

**Comment by frypf** [comment]
Been running for about 1 hour and so far not noticed any delays on showing the prefs window, which was happening quite frequently before.
If I quickly press & release to switch, or hold the shortcut to bring up the list then I've still noticed the delay intermittently, but seems much less frequent now. Attempting to retrace my steps, it seems this happened when the _previous_ switch triggered a space change, and then I call AltTab again shortly (within ~2s) after. If I hold the shortcut to look at the list then I just notice the delay in showing it, if I release before the switcher is shown then it manifests as the glitch from gif 3 above.

**Comment by lwouis** [comment]
I found another lead! A while ago, I implemented #127, and as part of that effort, I added a timer which checks every 5s if the system permissions (i.e. Accessibility and on Catalina, Screen Recording) are properly checked. It's unfortunate we have to use a timer, but there are no official API to listen for the Screen Recording changes, so there is no choice.

Turns out that at that time I put the timer on the main thread. This means that if for whatever reason, checking the permissions takes some time, it delays updating the UI. This timer is every 5s, so depending on system load and user action timing, it would explain these very random freezes of the UI we have been seeing.

I'll put the timer on a background thread. I think/hope that will finally mark the end of this performance journey!

**Comment by lwouis** [comment]
For archive purpose: I also investigated (again):

* App Nap
* Sudden Termination
* Automatic Termination

These features were not causing an issue there, but it's good to have that conclusion written here.

**Comment by mfn** [comment]
The build I started to use ( https://github.com/lwouis/alt-tab-macos/issues/563#issuecomment-688478090 ) does exhibit the latency before the AltTab windows appear.

Shall I upgrade to the latest official release now?

**Comment by mfn** [comment]
Delay is still happens at times on 6.3.0

**Comment by frypf** [comment]
Glad to hear you're narrowing it down. Managed to do some further narrowing myself for the issue in [gif 1](https://github.com/lwouis/alt-tab-macos/issues/563#issuecomment-688356181) after I'd noticed it happen a few more times since yesterday. I've only noticed this happen in fullscreen Chrome, and it seems to be related to significant changes in interactive screen content _underneath where the prefs window is due to appear_. (Eg. clicking web buttons, editing text). Again, this is not _wholly_ deterministic - it doesn't happen every time, but is much more likely. I've attempted to catch a gif highlighting the kind of content change that triggers it:

![prefsDelayContentChanged](https://user-images.githubusercontent.com/65832791/92491312-ff3bbb00-f1e9-11ea-84da-5fab9130cee3.gif)

It reminds me a lot of the issues you mention around [4:50 in this video](https://www.loom.com/share/a2b64f67ca074d869696a736012370df?t=290&autoplay=1). In fact, at 5:03 when you select some text in the GitHub comment box, it actually generates a bunch of accessibility events, perfectly illustrating your own point! Hope this is of use.

**Comment by frypf** [comment]
> Delay is still happens at times on 6.3.0

Yes but for me at least, it's orders of magnitude less than with 6.2.0, and seems to be related to either [WindowServer updates after changing space](https://github.com/lwouis/alt-tab-macos/issues/563#issuecomment-688499557), or possibly the circumstances described in my previous comment.

**Comment by frypf** [comment]
Bringing in @MrBaseMax's comment from #587:

> 2. And this menu behaves like the option "do nothing" turned on (stays and waits for a click). Which is not. (Also "sometimes").

I can consistently reproduce this one in these circumstances:

- Invoke AltTab shortcut causing a switch to an app in another space.
- Invoke the shortcut again during the space transition, releasing it before the transition completes _and_ before the "Apparition delay" time is complete.

**Comment by MrBaseMax** [comment]
> Bringing in @MrBaseMax's comment from #587:
> 
> > 1. And this menu behaves like the option "do nothing" turned on (stays and waits for a click). Which is not. (Also "sometimes").
> 
> I can consistently reproduce this one in these circumstances:
> 
> * Invoke AltTab shortcut causing a switch to an app in another space.
> * Invoke the shortcut again during the space transition, releasing it before the transition completes _and_ before the "Apparition delay" time is complete.

Yes, usually caused by the second shortcut pressed in a row. But i use mostly just a single space. So, do not see the dependency

**Comment by frypf** [comment]
> But i use mostly just a single space.

And no fullscreen apps?

**Comment by lwouis** [comment]
@frypf I was able to reproduce this issue. I will open another ticket for this. It's not related to freezes or visual glitches. It's a bug with shortcuts

**Comment by MrBaseMax** [comment]
6.4.0 negative. Still there

**Comment by MrBaseMax** [comment]
The key here is the time before release a shortcut. There's no bug when doing it slow

**Comment by frypf** [comment]
Lets continue the shortcut discussion in #588 that @lwouis just opened. Apologies for confusing that issue with this one.

**Comment by lwouis** [comment]
How about the freezes and visual glitches in 6.4.0?

**Comment by MrBaseMax** [comment]
> How about the freezes and visual glitches in 6.4.0?

Still there :(

**Comment by lwouis** [comment]
What exactly do you observe? We need detailed a description, and steps to reproduce and video would help a lot of course. Only through detailed feedback will we have a chance of assessing the current situation, and come up with a solution.

**Comment by frypf** [comment]
> How about the freezes and visual glitches in 6.4.0?

Still there for me but very intermittent and I'm increasingly sure only:

- shortly after a space change (switcher window delayed)
- shortly after computer wakes from sleep (switcher window delayed)
- after interacting with web content (prefs / about / update window delayed)<sup>[1]</sup>


[1] Now confirmed this separately in both Brave and Chrome, can confirm fullscreen status is not pertinent.

**Comment by MrBaseMax** [comment]
> What exactly do you observe? We need detailed a description, and steps to reproduce and video would help a lot of course. Only through detailed feedback will we have a chance of assessing the current situation, and come up with a solution.

I see delays of a second or more, even when UI have to gather thumbs only for current app windows. More often it happens when changing combination (press 10 times combination 2 and then combination 1)

**Comment by frypf** [comment]
Ok I've noticed the delay showing the switcher once now where a space transition definitely wasn't involved. However, my jerry-rigged menubar CPU temp monitor was indicating a spike in activity at the time. @lwouis do you know if it's possible (or helpful) to run Instruments on WindowServer? I've found the executable in the private Skylight framework but obviously as it's single-instance, it just stops instantly. I just thought I'd try to see if my AltTab delays were showing any corresponding high CPU activity in WindowServer to at least rule that out.

**Comment by lwouis** [comment]
@frypf Use Instruments with `System Trace` instead of the usual `Time Profiler`. It will not only sample AltTab, but also sample the system process, including `WindowServer`. You can [see the example I posted above](https://github.com/lwouis/alt-tab-macos/issues/563#issuecomment-682389648). There are then many indicators and much data that you can look at. It's all fairly low-level though, so may be hard to read if you're not proficient in how macOS works.

**Comment by frypf** [comment]
Will move onto System Trace now. I just tried it with the general Activity Monitor profiler and managed to trigger another same-space delay, and checking back in Instruments it corresponded with WindowServer at 57.8% CPU and AltTab at 5.3%.

If I'm honest it's mostly been your videos that have given me any clue how to look at this stuff anyway, and I've found them very helpful while trying to develop my own little CLI thing. I can read a thousand manuals, but nothing teaches me better than watching someone doing something and trying it for myself.

**Comment by mfn** [comment]
FWIF, also seeing delays in bringing up the AltTab window with 6.4.0

**Comment by frypf** [comment]
> so may be hard to read if you're not proficient in how macOS works.

Wow, you were not kidding 😂

**Comment by lwouis** [comment]
Here is a [build in which I have disabled the sync loading of the first 3 thumbnails](https://github.com/lwouis/alt-tab-macos/files/5190304/AltTab.app.zip). All thumbnails are loaded async, after initial display of the UI in this build.

Please try it out and let me know if it improves things somehow

**Comment by gquittet** [comment]
On 6.4.0 (even on the custom build above), I've got the delay of 1s before Alt-Tab is showing. 

I'm on a MacBook Pro 15 2019.

**Comment by frypf** [comment]
Still same issues as described before, but after running System Trace a few times my delays in showing the switcher correspond to a spike in WindowServer activity just beforehand (within 0.4s), followed by a big spike in kernel activity, followed by small a spike in AltTab at the point I hit the shortcut. I'm definitely not able to make much sense of it beyond that, and after saving a couple of 10s traces I realised they were 1.56gb and 1.42gb (I know there's a lot of info in there, but this seems huge).

Just having these files open causes resource issues on my low-spec machine, and actually seems to increase the frequency of the WindowServer spikes and therefore also the frequency of the delay happening.

Attempting to quantify it a bit better, with 6.2.0 I would say that for me the delay manifested ~ 1 in 20 interactions with AltTab. After the async App Badge code you pushed that reduced the frequency to ⅓-½ of that, if not lower. Subsequent builds haven't really made a dent in it any further. A screenshot of (what I've judged to be) the most obvious parts of the .trace docs:

![Screenshot 2020-09-08 at 20 39 14](https://user-images.githubusercontent.com/65832791/92522313-7129f980-f216-11ea-962d-e0b8503d8197.png)

![Screenshot 2020-09-08 at 20 52 43](https://user-images.githubusercontent.com/65832791/92522325-75eead80-f216-11ea-99ea-41725e0e5923.png)

Please let me know if you need me to pull anything else from them - I'd upload them but they're massive and possibly not even of use.

**Comment by lwouis** [comment]
@frypf thanks for sharing this analysis. It's hard to draw conclusions though, as you've seen for yourself.

@gquittet you're saying that on 5.3.0 the freezes are gone? From your perspective, 6.0.0 introduced these issues? 6.0.0 introduced loading the thumbnails *after* showing the UI. So clearly if you have 30 windows open, it's possible that even with the freeze before the UI shows, it would still be faster than on 5.3.0 where there is a natural wait to show the UI as the thumbnails are being updated before showing anything. Doesn't 5.3.0 feel slow in that regard?

**Comment by lwouis** [comment]
On my machine, I see 3 problems happening:

* After not using AltTab in a few minutes, when I trigger it next, there is a delay before it shows its UI. This one feels to me like it could be due to:
    * The OS putting low priority on the app. You can set QoS for threads and dispatch queues, but how does it work for the main thread? Is there a priority penalty being a background app?
    * The OS putting its RAM in a special state?
    * ~The OS putting the app in App Nap?~ (there is now code to prevent that in AltTab)
    * Some cache being expired, like the `NSView` `layer` being cleared, and AltTab has to reconstruct it?

* Randomly, when the UI is in the process of appearing or disappearing, it shows this visual glitch of showing the outline of the window, but the contents are not correct. It's kind of transparent, showing the contents underneath, while still showing the window's box/outline and shadows.

* When AltTab shows its UI, and the window has changed size, it shows the previous window for a few frames, then shows the correct window. A way to trigger this is to switch between the 2 default shortcuts, or set the preference to show different types of windows, then trigger AltTab again so there is less/more windows to be shown.

These 3 issues seem independent to me, but maybe there are not. Regarding actionability, I think issue 3 is reproducible semi-consistently, and I have ideas on what's at play. For issue 1 and 2, I'm at a complete loss, without any lead on what is happening.

**Comment by mfn** [comment]
> * After not using AltTab in a few minutes, when I trigger it next, there is a delay before it shows its UI.

FTR this is the issue I'm practically encountering all the time. In fact, very often in recent times. Also happens with the build from https://github.com/lwouis/alt-tab-macos/issues/563#issuecomment-689055871

**Comment by mfn** [comment]
I again went back to v5.3.0 and yes, it feels immediately different if one pays attention:
- I do not have this "exorbitant high" (in comparison!) delays bringing up AltTab
- when v6 brings up the windows without the delay, it is definitely much faster than v5

But the occasional long delay, which basically makes me have to wait before I can switch windows, makes me rather use v5.3.0 and live with that it generally comes up slightly slower but I don't experience these longer delays at all.

**Comment by gquittet** [comment]
We should use `git bisect` to find the exact part of the code that has a huge impact on the performance between 5.3.0 and 6.4.0.

I can do it but I have no idea on how to compile alt tab on my Mac 🤔

**Comment by lwouis** [comment]
> We should use git bisect to find the exact part of the code that has a huge impact on the performance between 5.3.0 and 6.4.0.

It may be useful to download 6.0.0 and see if it displays the issues. That would reduce the scope of the bisection, if it turns out it was already happening in 6.0.0

> I have no idea on how to compile alt tab on my Mac 🤔

It's extremely easy. I [documented the steps here](https://alt-tab-macos.netlify.app/contributing).

**Comment by mrc247-DE** [comment]
after updating from 6.2.0 to 6.4.0 I notice a big delay when, in my case, pressing command + tab.

It doesn't always happen, but it happens frequently and causes irritation.

in version 6.2.0 everything was still ok.

**Comment by mfn** [comment]
I think we're narrowing it down!

I tested with 6.0.0 a while and didn't see this hangs.

**Comment by lwouis** [comment]
First, let's try to be precise, as there are [multiple issues at play here](https://github.com/lwouis/alt-tab-macos/issues/563#issuecomment-689296016).

Regarding issue 1 in the link above, I just tested all versions myself, and narrowed the regression. It happened between 6.2.0 and 6.3.0. I'll try to find which exact code change brings the issue now

**Comment by zzamboni** [comment]
Glad to see the long discussion here. I've been thinking over the last few days that I was slowly going insane, since I've seen some of the behaviors described here, on and off, for a few days. I am pretty sure they started at or after v6.0.0.

There are two particular behaviors I have observed:

- Slight delay when pressing the shortcut before the thumbnails come up. I thought it was computing the thumbnails. This seems to happen more often when I have my external 4K monitor connected, but it also happens without it. Seems more often after a while without having triggered AltTab.
- A few times, when I have my external monitor connected, I have seen AltTab completely fail to appear. It still switches to a new window, but the thumbnails never appear on screen. This has only happened with the external monitor, and in those cases it seems necessary to restart AltTab to bring it back.

I'm on 6.4.0 now, and the first issue is still happening. I have not seen the second one in a while, but I will keep an eye for it.

@lwouis what would be useful information to collect when these issues happen, to help you diagnose them?

**Comment by lwouis** [comment]
Hey everyone, so I think I identified the issues with the 29eff03. I made some changes, and i'm confident that issue 1 is gone. Issue 2 is still there though. And issue 3, I think is gone but less sure.

Could you [please try this build out](https://github.com/lwouis/alt-tab-macos/files/5195589/AltTab.app.zip), and share here if it still exhibits any of the 3 issues?

**Comment by mindfulsource** [comment]
> Hey everyone, so I think I identified the issues with the [29eff03](https://github.com/lwouis/alt-tab-macos/commit/29eff03e7048b77908ea2c8d21dca4939b07b07f). I made some changes, and I think that issue 1 is gone. Issue 2 is still there though. And issue 3, I think is gone but less sure.
> 
> Could you [please try this build out](https://github.com/lwouis/alt-tab-macos/files/5195589/AltTab.app.zip), and share here your results regarding the 3 issues?

Just tried this build for an hour. I will speak to the issue that I've observed. Some observations / possible clues:

- Don't think you've amended 'Shortcut 2' like you have 'Shortcut 1'. Higher rate of the slowness issue.

- The bug rate seems to be less if you are on dGPU (and not iGPU). My hardware: 16-inch MBP.

- Possible clue: I think some apps are exacerbating / triggering the issue. E.g. Google Chrome. I spent a good 20 minutes alt-tabbing between TextEdit and Firefox with zero slowness. This might have to do with video graphics somehow.

My 2c: until this particular bug can be resolved fully, I'd vote for making it be sightly slower vs. instant lightning speed, if it means less 'super slow' instances.

**Comment by lwouis** [comment]
@mindfulsource I can't believe you are still getting issue 1. For me, running 6.4.0 locally, i just wait, and can get the freeze when I trigger AltTab after a few minutes. Then i run the build I shared, and this is gone every single time. I tested for a few hours and the issue is just no longer happening for me locally.

Can you precisely describe which issue you are referring to? You may not be talking about issue 1.

**Comment by zzamboni** [comment]
@lwouis With the latest build, the delay on AltTab appearing seems to be gone. I have not tested in my dual-monitor setup, I will report later when I have a chance to test.

**Comment by nfekete** [comment]
I'm also on 6.0.0, which I reverted to using after facing the issue of the long delays in a later version (6.3.0 I think). For me, in 6.0.0 long delays rarely happen, while on 6.3.0 it happened quite often. I'll try the new 6.5.0 release and report back.

**Comment by mckennajones** [comment]
I noticed AltTab was taking a while to appear, clicked "Check for updates..." and there was the bug fix for this issue. Cheers @lwouis and everyone else who contributes to this awesome app!

**Comment by nfekete** [comment]
6.5.0 feels nice and snappy again. This is going to be my baseline version from now on, after 6.0.0.

**Comment by zzamboni** [comment]
Same for me - I've been using the earlier build and now 6.5.0 for a while, also on my dual-monitor setup, and it seems fine. Thanks @lwouis for all your work on diagnosing and fixing this!

**Comment by mindfulsource** [comment]
I've updated to v6.5.0 but I'm sorry but I'm still getting it. Not as often (which is great), but yes periodically. (I'm on forced dGPU right now. If I switched to auto / iGPU most of the time, I think the issue would come back more.)

I am on a MacBook Pro (16-inch, 2019). Specs:

2.6 GHz 6-Core Intel Core i7
64 GB 2667 MHz DDR4
AMD Radeon Pro 5500M 8 GB
Intel UHD Graphics 630 1536 MB

It may be related to some graphics issues this model is having. MacRumors forums have us talking about them.

**Comment by mfn** [comment]
> Hey everyone, so I think I identified the issues with the [29eff03](https://github.com/lwouis/alt-tab-macos/commit/29eff03e7048b77908ea2c8d21dca4939b07b07f). I made some changes, and I think that issue 1 is gone. Issue 2 is still there though. And issue 3, I think is gone but less sure.
> 
> Could you [please try this build out](https://github.com/lwouis/alt-tab-macos/files/5195589/AltTab.app.zip), and share here your results regarding the 3 issues?

Can confirm since I run this build (1h?) I haven't seen 1), the initial hang before the window appears, once. And usually it would _not take longer than a few minutes.

🎉

**Comment by mindfulsource** [comment]
I keep on getting it. I think I want to go back to an earlier version like 5.0, see if it goes away. Would that be helpful @lwouis?

**Comment by mfn** [comment]
> Can confirm since I run this build (1h?) I haven't seen 1), the initial hang before the window appears, once. And usually it would _not take longer than a few minutes.

In the meantime I could see the hangs but they seem much less frequent. First time I saw it since I installed the new build was a few minutes ago when switching between "in-app" windows but since then also saw it when switching between all apps windows.

----
Like, I was writing above sentence and writing and writing and though I would test it, press cmd-tab and did experience the hang (without switching. Then tested it again, super fast. Then tried cmd-~ for in-app switching and immediately also had the hang.

After writing above sentence and not switching away anyhere:
- cmd-tab => almost instant all windwos, esc close
- cmd-~ => hang

**Comment by lwouis** [comment]
Yes, any kind of "it *never* hanged on that version" information is of course useful. I found the regression above, clearly during my local tests. Now if you guys still experience issue, the question is: what's the version where it didn't exist for you? For me, the regression is clearly between 6.2.0 and 6.3.0 but apparently it's not the same for all of you

**Comment by mindfulsource** [comment]
I just tried v6.0.0. It still has the hang issue. Very early, very quickly, 3 times in just a couple minutes.

What's the earliest version that *anyone* has reported the issue (of delay in entire AltTab showing)? I could try going one revision before that and reporting results.

**Comment by mfn** [comment]
> For me, the regression is clearly between 6.2.0 and 6.3.0 but apparently it's not the same for all of you

I'll chase down this one from my side. It may take some time because as experience has shown, running 1h might not be enough to attest if the problem still exists or not

**Comment by mindfulsource** [comment]
v6.0.0 was stubbornly delaying so often that I then desperately tried an earlier version, v5.3.0. It was different, but still noticeable delay - this time it was EVERY time. About half a second. Cmd-Tilde for per-program grid was normal fast, but not for the whole grid of 30 windows.

Then weird observation: I switched into my secondary 'TestPerson' macOS user (already logged in) and it had an AltTab still open that's v6.4.0. I quitted that other AltTab and completely logged out of the user and went back to my normal user session. Restarted v5.3.0. Still exact same symptom though. In fact, I one instance of the 'full' 1.5s delay.

In further desperation, I rebooted my Mac, without ticking "Reopen windows when logging back in".

(I also went back to v6.5.0.)

Upon reboot: AltTab is in my autostart so it's already activated. Zero delay...so far so good. I open my 6 chrome windows. I open four more windows across two apps. still good. I open 4 Excel windows. still good. I open 10 TextEdit documents. still good. Yet another browser's many-tabbed window. still excellent. Opened two Finder windows (one with 6 tabs, another with 2)...still good.

27 windows and everything's still instant. After 30 more mins of using it before posting this, everything good.

However, there were a few in the next hour up until now.

- 1 came after I was in a VNC client window. (Not sure if relevant / clue: is it when another application is 'watching' or accessing your hotkey somehow...) (I notice that if I'm in the VNC client, the OS inside the vnc is a bit cluncky with how it handles cmd-tabbing out of the VNC client back into another local macOS window. It activates its start menu even due to unfinished Cmd downpress.)

After that once instance, everything still perfect in this session for a good while.

And my VNC app doesn't reliably produce the delay after that either. (not right after, nor any times after.) It was one time, but before I have noticed it can tend to produce the delay. This seems complex.

- Dammit. then noticed a second one. Another app which I don't want to mention for privacy reasons, but it has nothing much to do with listening for the Cmd key AFAIK.

Then I worked for further 30 mins.

- Last documented 1.5s delay while alt-tabbing out of a Firefox Developer private window into a non-private Firefox Developer window. (A very not often used window in the session...is _that_ a clue? I doubt it. Complex graphics issue?)

**Summary: after full macOS reboot and in v6.5.0, after one hour of use, 1.5s delay occurring only 3-4 times. Not enough times for me to worry about, but hope that is somehow useful feedback.**

**Comment by lwouis** [comment]
> Restarted v5.3.0. Still exact same symptom though.

Did you *ever* not have this issue? I think everyone else said that the issue didn't exist on 5.3.0. Could you test previous versions until it stops happening?

**Comment by mindfulsource** [comment]
Absolutely. I only noticed this for no more than I'd say 24-48 hours before opening [this ticket](https://github.com/lwouis/alt-tab-macos/issues/595) (not seeing this one here).

I unfortunately suspect a hardware bug may be part of what's going on. The 16-inch MBP has more than one graphics issue now, both of which I often experience. :/

The timing has coordinated with the discovery of one of those bugs on my Mac in only the same period - last 24-48 hours.

**Comment by frypf** [comment]
With 6.5.0 I've not noticed [issue 3](https://github.com/lwouis/alt-tab-macos/issues/563#issuecomment-689296016) at all. (I hadn't actually noticed this previously either due to the way I have the shortcuts / lists / "Then release" behaviour set up, but I changed my prefs back to the default for testing and have definitely not been able to reproduce it so far).

Issue 2 is still there, but only if I hit the shortcut twice in fairly quick succession. Reinvoking the shortcut after releasing all keys, it sometimes manifests on the second release.

I've not seen issue 1 occur at all in 6.5.0 _exactly_ as described (ie on the first press after a period of AltTab inactivity). Only in the same circumstances as issue 2, although obviously the UI delay manifests on the 2nd press. This and issue 2 have occurred with both shortcuts / window lists and with both cross-space / same-space switches - but definitely much more frequently when switching between windows on different spaces. A preceding period of not using AltTab markedly increases the likelihood of either 1 or 2 happening on the second invocation of the shortcut.

If I realise it's taking too long for the UI to appear, releasing the shortcut still switches window instantly. In these circumstances holding the modifier and repeating the non-modifier part also behaves as expected, switching to windows further along the list instantly but without the convenience of being able to see the switcher window. For this reason, issue 2 _feels_ like an extension of issue 1 to me - it seems like it manifests differently according to whether the delayed switcher window appears before / at / after the point I release the shortcut.

I'll go back and test on previous versions now and report back.

**Comment by mfn** [comment]
6.2.0, such a strange thing.

I noticed incredible hangs before alttab for all apps would come up.

But this was so clearly different, hanging _every time, that I had the feeling _something_ else is going on.

Before I sent an email with Thunderbird and switched away. Apparently the small "sending email" popup, which is supposed to disappear, seemed to be hung and I was the culprit in this case for AltTab.

Once I closed/started AltTab it would not even open up any window at all, clearly it was _completely_ hanging due to Thunderbird.

Closing AltTab, even the native OSX app switcher, which _did show_ Thunderbird, could not "activate" it.

After Force Quitting Thunderbird, AltTab insta-worked.

So, this was a case of "hang before alltab window is shown" but which is rather unique and none of my previous reports referred to something like this.

**Comment by frypf** [comment]
I've been comparing 5.3.0 and 6.0.0 and there's definitely something going on between these builds. I ran 5.3.0 for a half hour without issue, then upon switching to 6.0.0 I started noticing delays immediately so decided to categorically document what I noticed each time:

- 14:41 - switcher UI delayed show, same-space, 1st invocation of AltTab after opening
- 14:41 - quit and reopened AltTab to check if immediately reproducible (answer: no)
- 14:49 - switcher UI delayed show, same-space, after 2x shortcut, AltTab successfully invoked several times since reopening
- 14:53 - switcher UI delayed hide, cross-space, obvious visual glitch, AltTab successfully invoked several times since previous delay followed by short period of AltTab inactivity
- 14:59 - switcher UI delayed hide, same-space, after system wake
- 15:00 - switcher UI delayed hide, same-space, AltTab not activated since previous delay
- 15:03 - switcher UI delayed hide, cross-space, obvious visual glitch, AltTab successfully invoked several times since previous delay
- 15:07 - switcher UI delayed hide, cross-space, obvious visual glitch, AltTab not activated since previous delay
- 15:09 - switcher UI delayed hide, cross-space, obvious visual glitch, AltTab not activated since previous delay
- 15:11 - switcher UI delayed hide, cross-space, obvious visual glitch, AltTab not activated since previous delay
- 15:11 - switcher UI delayed hide, cross-space, obvious visual glitch, AltTab not activated since previous delay
- after this point similar symptoms were reproducible almost every time after >~30s of AltTab inactivity, tying in with [issue 1](https://github.com/lwouis/alt-tab-macos/issues/563#issuecomment-689296016).

I've been back on 5.3.0 again since, again without issue.


OTHER NOTES / OBSERVATIONS:
- When the switcher UI is delayed hiding, the switch still occurs instantly but the UI hangs around for ~1s (as in the original thread title).
- While I only really took note of this behaviour and joined the thread after updating to 6.1.0, I know that the delayed _hide_ stopped from 6.2.0.
-  While my previous perception was that the delayed _show_ came in from 6.2.0, my findings above indicate otherwise. But the delayed hide was far more prevalent.
- The visual glitch (relating to issue 2) on borders / shadow / NSVisualEffectView.Material background is made _much_ more obvious by cross-space transitions, but this is not to say it isn't there (just less apparent) for same-space. Others may have different experience of this eg if they're using same-space windows with more contrasting appearances (all my Desktop 1 windows are generally dark).
- From my own point of view, the partial rework of keyboard shortcut handling and showing/hiding the switcher during space transitions in 6.2.0 increased the frequency of me _noticing_ the visual glitch upon hitting the shortcut 2x across spaces, probably for that exact reason.
- The nature of the visual glitch also changed slightly at this point. What I noticed in the test above is that the switcher window shows the correct content, but the material blur / shadows / borders are all still acting upon the previous window's content. The newer visual glitch observed since 6.2.0 still shows the shadows and border, but the window content now fails to show either the NSVisualEffectView or the window list. This results in a transparent background retaining the content from the previous window, but with suspiciously similar timing to the original issue from the thread title. This is what I'm thinking I should test next (possibly in a slightly less anal-retentive fashion).

**Comment by zzamboni** [comment]
One more confirmation from my side - I used 6.5.0 all day today in dual-monitor setup, and noticed zero delays or glitches. For me at least, all three issues seem to be completely gone 🎉

**Comment by frypf** [comment]
Managed to catch a couple of gifs of the visual glitch with the delayed hiding from 6.0.0, highlighting what I was saying about the NSVisualEffect's background (and corner mask) retaining the previous window's content:

![switcherDelayedHide2](https://user-images.githubusercontent.com/65832791/92780788-72386380-f39b-11ea-8dc1-cf0b05333399.gif)

![switcherDelayedHide](https://user-images.githubusercontent.com/65832791/92780818-7bc1cb80-f39b-11ea-9192-8be624709b50.gif)

Comparing with the following gif from 6.5.0 (but observable from 6.2.0 onward), aside from the presence of the window list and the blurred UI background they do bear striking a similarity, and the timings are broadly similar.

![650glitch](https://user-images.githubusercontent.com/65832791/92784393-8467d100-f39e-11ea-88f9-b0649ecaa828.gif)

Timings from start of window switch to disappearance of visual glitch:
- 6.0.0:
    - 0.98s (gif1)
    - 0.86s (gif2)
- 6.5.0
    - 0.71s (gif3)
    - 1.15s (not shown)
    - 0.80s (not shown)
    - 0.85s (not shown)
    - 0.76s (not shown)

There may well be nothing to this, just being thorough 🤔

**Comment by mfn** [comment]
I think after a day using 6.2.0 I can say with good confidence that I did not not the "alttab window appearing hang" (Problem 1).

Except the one time when actual an application was directly responsible for this, see https://github.com/lwouis/alt-tab-macos/issues/563#issuecomment-690330008 .

Moving to 6.3.0 now

**Comment by frypf** [comment]
> I think after a day using 6.2.0 I can say with good confidence that I did not not the "alttab window appearing hang" (Problem 1).

This doesn't tally with my own experience. I read this and opened up 6.2.0 and was able to cause a delayed UI show within the first minute:

- Open AltTab and switch back and forth a few times between different apps (no issues)
- Don't activate AltTab for about 30s
- Switch and switch back in quick succession
- UI delayed show, same-space (iTerm > Finder) or cross-space (Chrome > Finder)

@mfn what version were you running on [this comment](https://github.com/lwouis/alt-tab-macos/issues/563#issuecomment-688289867)? I was definitely on 6.2.0 at that point and reported the same symptoms as you. I'm very conscious of the fact that others seem to be having a totally different experience to mine, so I've used different combinations of apps and both a bluetooth and wired keyboard for testing. My steps above are not 100% reproducible every time, but it happens more often than not regardless of apps / spaces / keyboard. The only version where I've not experienced any UI delays at all is 5.3.0.

**Comment by mindfulsource** [comment]
> Could you test previous versions until it stops happening?

I overlooked that you asked this. I just 2 days worth of testing. It's now been happening rarely, but I've tested by downgrading as soon as I notice it happen for the first time in the current version:

v6.5.0: happened.
v6.0.0: happened. (Took an entire day though. rRght after text searching for something in a large TextEdit RTF file.)
v5.3.0: same crappy observation as my previous message. EVERY TIME, there's an annoying, it feels like 0.5-second delay. Let's move down...
v5.0.0: exact same as 5.3.0.

At that point, I assume going earlier might break my config somehow.

I've gone back to v6.5.0 and don't know what else to contribute. If mostly gone for most of us (which now includes me), maybe we have to live with it.

**Comment by lwouis** [comment]
There is a lot of info in this thread. I've read everything carefully. I think the only actionable thread to follow for me at the moment is:

Try to simulate a frozen app, then see if having an unresponsive app around can contribute to AltTab micro-freezing.

The code is supposed to ignore frozen apps and display the ui instantly, while processing all blocking calls from the environment. If they eventually succeed (e.g. Photoshop finishes launching after the long splash screen, or a frozen app becomes responsive again) then we update asynchronously (including the UI if it's still open). There may be a bug or an edge-case not accounted for in that logic here

**Comment by MrBaseMax** [comment]
> v6.5.0: happened.
> v6.0.0: happened. (Took an entire day though. Right after text searching for something in a large TextEdit RTF file.)
> v5.3.0: same observation as my previous message: EVERY TIME, there's an annoying, it feels like 0.5-second delay. Let's move down...
> v5.0.0: exact same as 5.3.0.

I use 4.19.0 – no bugs for me 😀

**Comment by mindfulsource** [comment]
> > v6.5.0: happened.
> > v6.0.0: happened. (Took an entire day though. Right after text searching for something in a large TextEdit RTF file.)
> > v5.3.0: same observation as my previous message: EVERY TIME, there's an annoying, it feels like 0.5-second delay. Let's move down...
> > v5.0.0: exact same as 5.3.0.
> 
> I use 4.19.0 – no bugs for me 😀

OK. For science, I just tested 4.19.0. Observation: main alt-tab exact same as 5.3.0 (always, every time, noticeably slower speed than newer versions), but per-app alt-tilde'ing was instant which is interesting. (7 Chrome windows...)

I'm now back to latest version as it's better to be instant speed 99% of the time than noticeably slower every time...

**Comment by mfn** [comment]
From my comment https://github.com/lwouis/alt-tab-macos/issues/563#issuecomment-691055290

Actually I never managed to move to 6.3.0 and am still on 6.2.0 and, in the meantime, experiences these hangs.

Reason I came here is that I just have an interesting case:
- phpstorm hung, the process does not respond
  ![image](https://user-images.githubusercontent.com/87493/93751322-34d6a000-fbfd-11ea-8691-39f6c4c4996d.png)
- **every** AltTab invocation hung noticeable, even when I only wanted to view in-app windows of _other_ application like firefox
  It was absolutely reproducible for a multiple minutes while I tested around
- and then… I came here to write the issue, re-confirm this behaviour and… it's gone
  - phpstorm still hangs
  - but now AltTab is not impacted anymore

This is crazy.

But it led me to think that "AltTab external factors", i.e. latency imposed from other application during querying the window states or whatnot, might be related to this.

**Comment by lwouis** [comment]
@mfn thank you for sharing. I still have to simulate an app hanging, as I outline in my previous message. Let's see what that brings.

> Actually I never managed to move to 6.3.0 and am still on 6.2.0

Are you choosing not to update, or do you face a bug/issue when clicking to update? The latest version is 6.6.0

**Comment by mfn** [comment]
Nah, I just didn't came around to doing it.

Just upgraded to 6.6.0 👍

**Comment by mfn** [comment]
On 6.6.0 and still experiencing these delays, so nothing changes.

---
I already noticed something else for a while, hard to describe. When I switched between in-app, most often noticed this between terminals, I saw that first only part of the terminal was re-drawn, other parts were not. After a small latency, the missing parts were drawn.

When I tried to capture this with a screenshot, I got this though:
![image](https://user-images.githubusercontent.com/87493/94349496-16c9cf00-0045-11eb-9d8e-f131eb1641c2.png)

Now, that ~black part~ white part [1] is something I did not see; so this must be specific to the screenshot software I use. But it seems to match the window AltTab is using and then it hit me: the parts of the terminal not being redrawn are the parts where AltTab window _would have_ appeared. Though I never saw this because when switching alternatively between two terminal windows (using iTerm2  here), I hardly wait for AltTab. I just press <kbd>cmd</kbd>-<kbd>~</kbd> and immediately release it.

Not sure of this is related or not :/

[1] I swear when I used the software to make the screenshot, this was all black but after I pasted it here, it appeared white 🤷‍♀️

**Comment by mindfulsource** [comment]
I really need to say, this delay issue has been getting worse on my own computer the last 2 weeks. It's now happening so, so, very often. It's becoming a drag. Not just noticeable, actually degrading my computing experience.

(I'm staying up to date with new releases.)

It even just happened now when trying cmd-tilde on Notes.app, with only one Notes window open. The delay still happened, even just to bring up only one grid item.

**Comment by erentantekin** [comment]
This is happening for me quite often (6.12.0)

**Comment by mmattbtw** [comment]
I just started recieving this issue after updating my macOS version. I'm am on macOS 11.5.2 right now, I'm not sure what I was on before.

**Comment by 662** [comment]
I just started recieving this issue after updating my macOS version. I'm am on macOS 11.5.2 right now, I'm not sure what I was on before.

**Comment by lwouis** [comment]
@662 did you just post the same exact comment as [the one above yours](https://github.com/lwouis/alt-tab-macos/issues/563#issuecomment-913258275)? 🤔

**Comment by 662** [comment]
@lwouis Yes, the same issue  as @mmattbtw.

**Comment by erentantekin** [comment]
I used to experience this issue a lot with older versions. For me, 6.22.0 doesn't have this issue at all. But when I tried to update, I noticed the issue comes back so I downgraded back to 6.22.0.

FWIW, I'd recommend trying 6.22.0 for those experiencing this issue.

**Comment by lwouis** [comment]
The code that I suspect to be the reason for the more frequent issue is unfortunately accessibility code I added because a user explained that they lost sight and would like AltTab to support VoiceOver. I was touched by their situation and message and implemented VoiceOver.

Now for some reason that's beyond my understanding, the seemingly unrelated accessibility code seem to interfere and trigger this ticket issue. I implemented a workaround to mitigate it in v6.24, but it seems to still be an issue for some people. If you're familiar with macOS development, please [take a look here](https://github.com/lwouis/alt-tab-macos/blob/8ab0e615c84cbba820d6e7643bb06353cc4505d9/src/logic/Windows.swift#L105) and let me know if you get a sense of what could be at play. The delays are happening within apple frameworks. I don't know if it's a bug on their side, or a setup of our side that generates the delays on their side. Really hard to tell. Apple docs are poor and I have no other project to compare with, or expert macOS developer to ask advise from.

**Comment by unikzforce** [comment]
Hi guys! thanks a lot for this beautiful and ultra useful app.

I'm not familiar with the code base. so to understand how AltTab works i just played a youtube video and while It was playing I just did the AltTab. And understood that AltTab is taking an instant picture from each window instantly at the moment it is being initiated.

Isn't it possible that instead of using an on demand mechanism, use a background loop that do this job from time to time in a hybrid manner?

The idea is that AltTab can maintain a List of window captures during it's lifecycle.

And this list could be updated upon these circumestances:

1. when capture List is empty and user just presses AltTab --> the list must be initiated
2. Every N seconds the capture list could be updated, which N can be dynamically adjusted in a trade of between quality and performance, and can have a minimum and maximum amount.
3. Upon each important window event, the corresponding item in the capture list could be deleted or updated, or a new item could be added.

And it could be disabled by default.

for example if a user is looking at his VS Code for 1 hour, and not switching to anything else, that number N could gradully become 100 Seconds ( these numbers are just examples).

**Comment by unikzforce** [comment]
@erentantekin As I played that youtube movie, I understood that HyperSwitch doesn't seem to update the captured screen at all, It would capture a picture from the a window some moments after window is initiated. It doesn't update it at all it seems.

**Comment by unikzforce** [comment]
My bad, It seems that you have already implemented Async

**Comment by lwouis** [comment]
@unikzforce AltTab does a mix: after being invoked, it synchronously gets the first 3 screenshots, then display with these 3 fresh + old cached screenshots for other windows, then asynchronously gets the other screenshots and replaces them one by one. It all happens very fast usually so users may not even notice part of it is async.

That being said, the recent update made the delay in this issue happen more, and is unrelated to screenshots. I suspect thus that this delay is due to another cause. Please see my [explanations above](https://github.com/lwouis/alt-tab-macos/issues/563#issuecomment-932972974)

**Comment by davidhopkins** [comment]
> I've downgraded to 6.22.1 - Much faster and consistent. Unfortunately I miss out on some of the newer fixes.

How did you downgrade? I've tried install the specific version with brew but that seems to be a whole can of worms.

**Comment by Pierce913** [comment]
> > I've downgraded to 6.22.1 - Much faster and consistent. Unfortunately I miss out on some of the newer fixes.
> 
> How did you downgrade? I tried installing a specific version with brew but that seems to be a whole can of worms.

Just downloaded the zip from the releases section. Deleted existing version and installed new one.

**Comment by zacharee** [comment]
I played around with this a bit, and I think it might actually be related to handling Accessibility Events. I wrapped the logic in `handleEvent()` in an async and the freezes are at least less frequent. Also running the window focus logic async helps.

https://github.com/zacharee/alt-tab-macos/commit/194bc849b311720f6112064694193144f71237bb
https://github.com/zacharee/alt-tab-macos/commit/89811ff5fb559bb13c5d97e875e865f917cd03ea

**Comment by lwouis** [comment]
@zacharee thank you for trying to fix that issue. In my mind, it's the last issue i would like to fix before i would consider AltTab "done". Once this is fixed, everything else is new features, quality of life, and working arounds other apps bugs. This is the only issue i consider a significant bug in AltTab as it seems to impact many people, and in various ways.

What's your methodology to measure progress here? Me I can't reproduce any of the issue people have reported in this thread in a consistent way.

Your first commit brings issues i think as the events are processed on a background thread for a reason. Moving to the main thread will probably bring bugs and performance issue if these events are slow to be processed.

The second commit may be interesting but i'll need to be at my PC to confirm.

So far the only clue i have to locate the issue is that people are clear that there was a regression in that version: https://github.com/lwouis/alt-tab-macos/issues/563#issuecomment-932972974

That leaves only a few lines of code that somehow increase the occurrences of the issue. This should help us locate the issue, but i couldn't solve the mystery though. Your idea that the main thread is blocked is interesting but we need more understand: what is it blocked on? If you can reproduce the issue, then you could use Instruments and profile the app.

Note that i've dealt a lot with something in the past: AltTab blocking on an API call because macOS is slow (multiple seconds) to return the value. Of course Apple doesn't document any of this. This is the reason why accessibility events are processed in a concurrent queue on a background thread actually: this could block the main thread when macOS's WindowServer is busy for example, and the API returns slowly; But even off the main thread events can block each other. We had the case of Gimp being slow to open with a loading splash screen. The OS API would then block until Gimp is ready. We need thus to handle these events concurrently so that one doesn't block all subsequent ones.

**Comment by zacharee** [comment]
The problem is that it's not consistent to reproduce. The best way I've found is to have something intense like a Zoom meeting running and just spam the switch shortcut until it happens (or doesn't happen). That's partly why I think it's related to some sort of Thread blocking, since it only seems to happen when there's higher CPU/RAM usage.

It's definitely possible the actual cause is yet another framework bug, but it seems like it might be mitigated by running more things concurrently. The changes I made definitely aren't a full solution or workaround, though. The freeze happens less often, but it still happens.

**Comment by zacharee** [comment]
So I was looking into when the `voiceOverFocusedWindow()` function was getting called and noticed it was being called twice at the same time when the switcher was shown (and successfully passing the guard both times). I traced it back to the `handleEvent()` function and noticed that the conditional for allowing accessibility events to execute was this:

    !(pid == ProcessInfo.processInfo.processIdentifier && element.subrole() == kAXUnknownSubrole)

I think it should instead be this:

    !(pid == ProcessInfo.processInfo.processIdentifier || element.subrole() == kAXUnknownSubrole)

If I'm reading it right, the events should only be acted upon if both the PID != alt-tab's PID *and* the subrole != `kAXUnknownSubrole`. But the && meant it was running if *either* one of those conditions passed.

I'm going to keep testing, but I think this helped out at least a bit, since the `refreshOpenUi()` function isn't being called twice now.

**Comment by DavithkbY** [comment]
I can reproduce this in second by just alt tabbing fast, like I usually do. Now my alttab is even broken and frozen.
I can't alttab anymore and I get a circle turning icon on my cursor when I hover over the alttab application. It's just stuck like this:
<img width="1728" alt="image" src="https://user-images.githubusercontent.com/71630103/140745257-29d19bd4-1fbd-45f5-87f7-3eee95830231.png">

**Comment by DavithkbY** [comment]
video: https://streamable.com/pyptiu https://streamable.com/18ne7e , it happens almost every few minutes of normal use.

**Comment by zacharee** [comment]
Here's a test build that should fix the freezing. It's possible it's ARM-only, since I'm building on an M1 Mac, and it's got some other unrelated changes (like a fade-in animation and different bundle ID).

You might also need to run `sudo xattr -cr /path/to/AltTab.app` before macOS will let you run it.

@DavithkbY and @mfn would you be willing to try this build to see if it fixes the freezing issue for you? Make sure the original AltTab process is killed before running this one.

[AltTab.zip](https://github.com/lwouis/alt-tab-macos/files/7498346/AltTab.zip)

**Comment by DavithkbY** [comment]
> sudo xattr -cr

I think you found the gold at the end of the rainbow. It seems to have fixed it. If I experience any more lagg, I'll make sure to let you know, thanks!

**Comment by FinallyFinancialFreedom** [comment]
> sudo xattr -cr /path/to/AltTab.app

this app require macos 11, sudo xattr -cr, Where to Get a Catalina version?

**Comment by FinallyFinancialFreedom** [comment]
> Here's a test build that should fix the freezing. It's possible it's ARM-only, since I'm building on an M1 Mac, and it's got some other unrelated changes (like a fade-in animation and different bundle ID).
> 
> You might also need to run `sudo xattr -cr /path/to/AltTab.app` before macOS will let you run it.
> 
> @DavithkbY and @mfn would you be willing to try this build to see if it fixes the freezing issue for you? Make sure the original AltTab process is killed before running this one.
> 
> [AltTab.zip](https://github.com/lwouis/alt-tab-macos/files/7498346/AltTab.zip)

this app require macos 11, sudo xattr -cr, Where to Get a Catalina version?

**Comment by zacharee** [comment]
This should be a universal binary for 10.12 and later @FinallyFinancialFreedom.
[AltTab.zip](https://github.com/lwouis/alt-tab-macos/files/7505322/AltTab.zip)

**Comment by zacharee** [comment]
> @zacharee I think you've done it! I had been following this thread since its beginning since it was an issue I faced every day every few minutes, but your (first) build has been working flawlessly on Big Sur with no visual hangs or delays for at least 2 hours for me.
> 
> I took a peek at the PR out of interest but I unfortunately couldn't think of why the change fixes the problem. :(

That's good to hear!

From what I could tell, refreshOpenUi was being called twice for every focus change, which must have caused some sort of deadlock situation. Making sure that alt-tab couldn't trigger itself from a focus change accessibility event meant refreshOpenUi would only run once.

**Comment by FinallyFinancialFreedom** [comment]
> This should be a universal binary for 10.12 and later @FinallyFinancialFreedom. [AltTab.zip](https://github.com/lwouis/alt-tab-macos/files/7505322/AltTab.zip)

Amazing! It solves all my problems, main window show up delay, fade delay. response slow. thx.

**Comment by zacharee** [comment]
@FinallyFinancialFreedom, @yutotakano, @DavithkbY, can you try this build, just to make sure there aren't any regressions? It performs fine for me so far.

[AltTab.zip](https://github.com/lwouis/alt-tab-macos/files/7513359/AltTab.zip)

**Comment by zacharee** [comment]
> Whenever I click on the "Appearance" tab it crashes the app, this wasn't the case with previous builds, aside from that, the lag itself is gone, so great work on the lag thingy! Here is the crash log from when I click the Appearance tab [crash.log](https://github.com/lwouis/alt-tab-macos/files/7513647/crash.log)

Hmm. I'm not getting a crash. Are you on x86? Maybe it's some effect of the universal binary. I know the project isn't quite ready to be built like that yet.

**Comment by zzamboni** [comment]
@zacharee I've been running your build from yesterday with no issues, and I just updated to the one you posted today. Both have been working fine so far, without some of the glitches and delays I had been experiencing before. Thanks!

**Comment by destronije** [comment]
> > Whenever I click on the "Appearance" tab it crashes the app, this wasn't the case with previous builds, aside from that, the lag itself is gone, so great work on the lag thingy! Here is the crash log from when I click the Appearance tab [crash.log](https://github.com/lwouis/alt-tab-macos/files/7513647/crash.log)
> 
> Hmm. I'm not getting a crash. Are you on x86? Maybe it's some effect of the universal binary. I know the project isn't quite ready to be built like that yet.

I'm using MacBook Pro 2019 with I7
Yep, it might be because of the universal binary

**Comment by DavithkbY** [comment]
@zacharee build of yesterday and today perform the same for me, flawlessly. M1 PRO Monterey 12.0.1

**Comment by zacharee** [comment]
One more build to try. @DavithkbY, @destronije, @zzamboni, @FinallyFinancialFreedom, @yutotakano.

[AltTab.zip](https://github.com/lwouis/alt-tab-macos/files/7520576/AltTab.zip)

**Comment by DavithkbY** [comment]
> One more build to try. @DavithkbY, @destronije, @zzamboni, @FinallyFinancialFreedom, @yutotakano.
> 
> [AltTab.zip](https://github.com/lwouis/alt-tab-macos/files/7520576/AltTab.zip)

still works very well for me :)

**Comment by yutotakano** [comment]
Same here, tried the [latest build](https://github.com/lwouis/alt-tab-macos/issues/563#issuecomment-966285500) for four hours now, works without problems. Thanks!

**Comment by destronije** [comment]
> One more build to try. @DavithkbY, @destronije, @zzamboni, @FinallyFinancialFreedom, @yutotakano.
> 
> [AltTab.zip](https://github.com/lwouis/alt-tab-macos/files/7520576/AltTab.zip)

Perfect, don't touch anything :)

**Comment by FinallyFinancialFreedom** [comment]
> One more build to try. @DavithkbY, @destronije, @zzamboni, @FinallyFinancialFreedom, @yutotakano.
> 
> [AltTab.zip](https://github.com/lwouis/alt-tab-macos/files/7520576/AltTab.zip)
@zacharee  Every things is fine. thx
A little problem, the position of the switch main window now show up in the very middle height of the screen, although previous version show in the more higher than now. Can i adjust it?

**Comment by zacharee** [comment]
> > One more build to try. @DavithkbY, @destronije, @zzamboni, @FinallyFinancialFreedom, @yutotakano.
> > [AltTab.zip](https://github.com/lwouis/alt-tab-macos/files/7520576/AltTab.zip)
> 
> @zacharee Every things is fine. thx A little problem, the position of the switch main window now show up in the very middle height of the screen, although previous version show in the more higher than now. Can i adjust it?

The builds I've been sharing are based on my fork's master branch, where I've made changes like adjusting the vertical alignment. @lwouis' official builds will have the normal alignment.

**Comment by theicfire** [comment]
OMG YAY!!!!!! It works so well 😭 😭 😭
